### PR TITLE
Directly wire ctx into http request

### PR DESCRIPTION
**Pull request by rht**
closes #1993

### CODE REVIEW

**Comment by whyrusleeping** [comment]
drop the goroutine that the 'Do' is in entirely, and remove the select channel stuff, see https://github.com/ipfs/go-ipfs/commit/590e1c86b4ae3e44460f9e2c7bf0e6b8956cbe4a

**Comment by whyrusleeping** [comment]
@rht i'm out for the day, but if you can figure out how to get CloseNotify working again, I would be soooooo happy

File: commands/http/client.go [hunk 1]
```diff
@@ -30,20 +30,13 @@ type Client interface {
 
 type client struct {
 	serverAddress string
-	httpClient    http.Client
+	httpClient    *http.Client
 }
 
 func NewClient(address string) Client {
-	// We cannot use the default transport because of a bug in go's connection reuse
```
Comment (rht) [inline commands/http/client.go]: How to check against this case?

Comment (whyrusleeping) [inline commands/http/client.go]: this comment can be removed. We're good now that we've switched to go1.5

File: cmd/ipfs/main.go [hunk 2]
```diff
@@ -685,3 +685,10 @@ func isConnRefused(err error) bool {
 
 	return netoperr.Op == "dial"
 }
+
+func wrapContextCanceled(err error) error {
+	if strings.Contains(err.Error(), "request canceled") {
+		err = errors.New("request canceled")
```
Comment (whyrusleeping) [inline cmd/ipfs/main.go]: I would go ahead and turn this into `ErrRequestCancelled` or something.

File: cmd/ipfs/main.go [hunk 3]
```diff
@@ -39,6 +39,7 @@ var log = logging.Logger("cmd/ipfs")
 var (
 	errUnexpectedApiOutput = errors.New("api returned unexpected output")
 	errApiVersionMismatch  = errors.New("api version mismatch")
+	errRequestCanceled     = errors.New("request canceled")
```
Comment (rht) [inline cmd/ipfs/main.go]: err defined. I will make it exportable once I've found the place that needs the ctx-cancel-wrapped.

File: commands/http/client.go [hunk 4]
```diff
@@ -30,20 +30,13 @@ type Client interface {
...
-	// code. It causes random failures in the connection including io.EOF and connection
-	// refused on 'client.Do'
 	return &client{
 		serverAddress: address,
-		httpClient: http.Client{
-			Transport: &http.Transport{
-				DisableKeepAlives: true,
```
Comment (jbenet) [inline commands/http/client.go]: do we still want to disable keep alives or no?

File: commands/http/client.go [hunk 5]
```diff
@@ -103,46 +96,27 @@ func (c *client) Send(req cmds.Request) (cmds.Response, error) {
 	version := config.CurrentVersionNumber
 	httpReq.Header.Set(uaHeader, fmt.Sprintf("/go-ipfs/%s/", version))
 
-	ec := make(chan error, 1)
-	rc := make(chan cmds.Response, 1)
-	dc := req.Context().Done()
+	httpReq.Cancel = req.Context().Done()
```
Comment (jbenet) [inline commands/http/client.go]: oh that's nice

Comment (jbenet) [inline commands/http/client.go]: can we have a test here to make sure this works as we expect it to? like have some request hang and have a context expire + be cancelled out? (doesnt have to be sharness, could be golang test)

**Comment by jbenet** [comment]
this LGTM in general. but want a test to make sure this behaves as we expect it to-- i've been misled by interpretations of standard libs before.

**Comment by rht** [comment]
Though in this case, using the stdlib prevents a class of race conditions.
Test is possible once #2032 is merged.

**Comment by whyrusleeping** [comment]
@rht we've got `CloseNotify` fixed now, mind rebasing this one on master?

**Comment by rht** [comment]
This could have been merged with/without the `CloseNotify`.

**Pull request closed without merging**